src/highlight/HighlightList.tsx

```diff
@@ -1,20 +1,62 @@
 import React from 'react';
 import classNames from 'classnames';
-import HighlightAnnotation from './HighlightAnnotation';
+import HighlightCanvas, { CanvasShape } from './HighlightCanvas';
+import HighlightSvg from './HighlightSvg';
+import HighlightTarget from './HighlightTarget';
 import useOutsideEvent from '../common/useOutsideEvent';
-import { AnnotationHighlight } from '../@types';
-
+import { AnnotationHighlight, Rect, TargetHighlight } from '../@types';
+import { checkValue } from '../utils/util';
 import './HighlightList.scss';
 
 export type Props = {
+    activeId?: string | null;
     annotations: AnnotationHighlight[];
     className?: string;
     onSelect?: (annotationId: string | null) => void;
 };
 
-export function HighlightList({ annotations, className, onSelect }: Props): JSX.Element {
+export function filterHighlight({ target }: { target: TargetHighlight }): boolean {
+    const { shapes = [] } = target;
+
+    return shapes.every(
+        ({ height, width, x, y }: Rect) => checkValue(height) && checkValue(width) && checkValue(x) && checkValue(y),
+    );
+}
+
+export function getHighlightArea(shapes: Rect[]): number {
+    return shapes.reduce((area, { height, width }) => area + height * width, 0);
+}
+
+export function sortHighlight(
+    { target: targetA }: { target: TargetHighlight },
+    { target: targetB }: { target: TargetHighlight },
+): number {
+    const { shapes: shapesA } = targetA;
+    const { shapes: shapesB } = targetB;
+
+    // Render the smallest highlights last to ensure they are always clickable
+    return getHighlightArea(shapesA) > getHighlightArea(shapesB) ? -1 : 1;
+}
+
+export function getHighlightShapesFromAnnotations(
+    annotations: AnnotationHighlight[],
+    activeId: string | null,
+): CanvasShape[] {
+    return annotations.reduce<CanvasShape[]>((currentShapes, annotation) => {
+        const {
+            id,
+            target: { shapes },
+        } = annotation;
+
+        return currentShapes.concat(shapes.map(shape => ({ ...shape, isActive: activeId === id })));
+    }, []);
+}
+
+export function HighlightList({ activeId = null, annotations, className, onSelect }: Props): JSX.Element {
     const [isListening, setIsListening] = React.useState(true);
     const rootElRef = React.createRef<SVGSVGElement>();
```

Comment:
Minor, but we should probably rename this to `svgElRef` or something, since it's not the root element.